### ISSUE TITLE
[RaySGD] How workers sync is handled?

### ISSUE THREAD

**Issue by marcociccone** [ray-project/ray#7624]
Hi, I started using Ray (Tune) together with my labmates (great tool, thanks!) and we wanted to build an infrastructure based on it to allow our students to run experiments and scale up easily. 
We noticed that you were developing RaySGD so we were thinking to take inspiration from it or maybe even adopt it as is.

We are using PyTorch, however, there is something that we don't understand how workers interact with each other: 
- Workers seem to be parallel trainers that consume batches after the have been split. Each worker may be using a gpu or not and train an epoch independently from the others. You are using `DistributedDataParallel` (ddp), [but it seems that each worker has its own ddp](https://github.com/ray-project/ray/blob/45ce40e5d44801193220d2c546be8de0feeef988/python/ray/util/sgd/torch/torch_trainer.py#L279-L292) and optimizer, so I don't understand where the reduction happens. 

Which is the trick?

Thanks!

**Comment by richardliaw** [comment]
DistributedDataParallel handles the synchronization/reduction. Maybe this will help! https://datascience.stackexchange.com/questions/22191/what-is-the-difference-between-pytorchs-dataparallel-and-distributeddataparalle

**Comment by marcociccone** [comment]
Thanks, I figured that, but shouldn't be a single Data parallel table ? From what I see there is one for each replica. Am I missing something?

**Comment by richardliaw** [comment]
if by table you mean data, RaySGD automatically uses https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler
to make sure every iterator has a new subset of the data.

**Comment by marcociccone** [comment]
Oh no, (I don't know where table came from) I was still referring to the DistributedDataParallel. I mean, aren't you instantiating a new DistributedTorchRunner and hence a DsitrbutedDataParallel for each worker [here](https://github.com/ray-project/ray/blob/45ce40e5d44801193220d2c546be8de0feeef988/python/ray/util/sgd/torch/torch_trainer.py#L279-L292), when you call `Runner.remote(...)`? 

I guess what I am trying to understand is if each worker has its own DistributedDataParallel, Optimizer etc.. and if so why? Shouldn't be only one optimizer?

**Comment by richardliaw** [comment]
DistributedDataParallel follows the following algorithm:

1. Run multiple copies of the training script and each copy:
  1. reads a chunk of the data
  2. runs it through the model
  3. computes model updates (gradients)
2. Average gradients among those multiple copies
3. Update the model (on all copies)
4. Repeat (from Step 1)

Note that the model updating is across all workers. If the model is only updated on one node, you'll need to ship the model out to all workers again, after each batch.

This is the same algorithm used in Horovod, Baidu Allreduce, etc.
https://towardsdatascience.com/distributed-tensorflow-using-horovod-6d572f8790c4

**Comment by richardliaw** [comment]
(let me know if you have any other questions)

**Comment by richardliaw** [comment]
closing this for now, feel free to reopen later.
